File: data-structure-algorithm/83-Remove_Duplicates_from_Sorted_List.js
/**
 * 
 * 
 * Given a sorted linked list, delete all duplicates such that each element appear only once.

  Example 1:

  Input: 1->1->2
  Output: 1->2
  Example 2:

  Input: 1->1->2->3->3
  Output: 1->2->3
 * 
 * 
 */

 /**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
   if (head == null) return null
   let curHead = head
   let cur = head.val // 这个变量实际上不需要单独存储
   let node = head
   while(node != null) {
    let nextNode = node.next
    if (nextNode != null) {
      if(nextNode.val == cur) {
        node.next = nextNode.next
      } else {
        cur = nextNode.val
        node = node.next
      }
    } else {
        break // 这里break的话， 可以直接将node.next 和node 不为null都放在while的判定条件里面
    }
   }
   return curHead
};


// much better
var deleteDuplicates = function(head) {
  let node = head
  while (node && node.next) {
    if (node.val === node.next.val) {
      node.next = node.next.next
    } else {
      node = node.next
    }
  }
  return head
}

File: data-structure-algorithm/golang/83. Remove Duplicates from Sorted List.go
package ds

/**
Given the head of a sorted linked list,

delete all duplicates such that each element appears only once. Return the linked list sorted as well.


The number of nodes in the list is in the range [0, 300].
-100 <= Node.val <= 100
The list is guaranteed to be sorted in ascending order.

**/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

// 方法一，数组存，然后去重复，然后重新构造一个链表

// 方法二, 一次线性遍历，画个图更好理解
func deleteDuplicates(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	h := &ListNode{}
	h.Next = head
	pre := h

	for h.Next != nil && h.Next.Next != nil {
		for h.Next != nil && h.Next.Next != nil && h.Next.Val == h.Next.Next.Val {
			h.Next = h.Next.Next
		}
		h = h.Next
	}

	return pre.Next
}

// 对方法二的改进，去掉pre节点，better

func deleteDuplicates2(head *ListNode) *ListNode {
	h := head
	for h != nil && h.Next != nil {
		if h.Val == h.Next.Val {
			h.Next = h.Next.Next
		} else {
			h = h.Next
		}
	}
	return head
}
